core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java

```diff
@@ -3167,6 +3172,412 @@ public void testRegisterExistingTable() {
+        .isInstanceOf(IllegalArgumentException.class)
+        .hasMessageContaining("Tags do not support setting minSnapshotsToKeep");
+  }
+
+  @Test
+  public void testMultiReferenceManagement() {
+    C catalog = catalog();
+
+    if (requiresNamespaceCreate()) {
+      catalog.createNamespace(NS);
+    }
+
+    Table table = catalog.buildTable(TABLE, SCHEMA).create();
+    table.newFastAppend().appendFile(FILE_A).commit();
+    Snapshot baseSnapshot = table.currentSnapshot();
+
+    String branch1 = "feature-1";
+    String branch2 = "feature-2";
+    String branch3 = "hotfix";
+    String tag1 = "v1.0";
+    String tag2 = "v1.1";
+    String tag3 = "release-candidate";
+
+    table
+        .manageSnapshots()
+        .createBranch(branch1, baseSnapshot.snapshotId())
+        .createBranch(branch2, baseSnapshot.snapshotId())
+        .createBranch(branch3, baseSnapshot.snapshotId())
+        .createTag(tag1, baseSnapshot.snapshotId())
+        .createTag(tag2, baseSnapshot.snapshotId())
+        .createTag(tag3, baseSnapshot.snapshotId())
+        .commit();
+
+    // Verify all branches and tags exist
+    Map<String, SnapshotRef> refs = table.refs();
+    assertThat(refs).containsKey(branch1);
+    assertThat(refs).containsKey(branch2);
+    assertThat(refs).containsKey(branch3);
+    assertThat(refs).containsKey(tag1);
+    assertThat(refs).containsKey(tag2);
+    assertThat(refs).containsKey(tag3);
+
+    // Verify types
+    assertThat(refs.get(branch1).isBranch()).isTrue();
+    assertThat(refs.get(branch2).isBranch()).isTrue();
+    assertThat(refs.get(branch3).isBranch()).isTrue();
+    assertThat(refs.get(tag1).isTag()).isTrue();
+    assertThat(refs.get(tag2).isTag()).isTrue();
+    assertThat(refs.get(tag3).isTag()).isTrue();
+
+    // Add data to main branch
+    table.newFastAppend().appendFile(FILE_B).commit();
+    Snapshot mainSnapshot2 = table.currentSnapshot();
+
+    // Add data to feature branch
+    table.newFastAppend().toBranch(branch1).appendFile(FILE_C).commit();
+    Snapshot branchSnapshot = table.snapshot(branch1);
+
+    // Verify isolation
+    refs = table.refs();
+    SnapshotRef mainRef = refs.get(SnapshotRef.MAIN_BRANCH);
+    SnapshotRef branchRef = refs.get(branch1);
+    SnapshotRef tagRef = refs.get(tag1);
+
+    // Main branch should point to latest main snapshot
+    assertThat(mainRef.snapshotId()).isEqualTo(mainSnapshot2.snapshotId());
+
+    // Feature branch should point to its own snapshot
+    assertThat(branchRef.snapshotId()).isEqualTo(branchSnapshot.snapshotId());
+    assertThat(branchRef.snapshotId()).isNotEqualTo(mainRef.snapshotId());
+
+    // Tag should still point to original snapshot
+    assertThat(tagRef.snapshotId()).isEqualTo(baseSnapshot.snapshotId());
+    assertThat(tagRef.snapshotId()).isNotEqualTo(mainRef.snapshotId());
+    assertThat(tagRef.snapshotId()).isNotEqualTo(branchRef.snapshotId());
```

Comment:
Nit: I feel like these 2 assertions aren't strictly neccessary for the abstraction we're testing